lib/OpenQA/Script/CloneJob.pm

```diff
@@ -226,11 +226,31 @@ sub handle_tx ($tx, $url_handler, $options, $jobs) {
     }
 }
 
+# append a formatted "-NN" to the TEST parameter
+sub append_idx_to_test_name ($n, $post_params) {
+    foreach my $job_key (keys %$post_params) {
+        my $job = $post_params->{$job_key};
+        my $test_name = $job->{TEST};
+        if ($n == 1) {
+            $test_name .= '-1';    # just append at the first time
+        }
+        else {
+            # from the second onwards, replace with the number
+            $test_name =~ s/-\d+$/\-$n/;
+        }
+        $job->{TEST} = $test_name;
```

Comment:
```suggestion
        # just append at the first time. From the second onwards, replace with the number
        $job->{TEST} = $n == 1 ? '-1' : $job->{TEST} =~ s/-\d+$/\-$n/r;
```

Reply:
Won't this replace the whole test name with `-1` on first pass ? I think we should keep the original name and just append the number

Reply:
ah, yes of course. Corrected my suggestion

Reply:
Honestly, I am a bit against this kind of oneliners. I know it's cool to "compress" 9 lines in one, but you gain nothing in terms of performance or feature, and just make the code more difficult to understand for a contributor. I'm in favor of clean, expressive code, even if it's Perl :)

Reply:
This is not about oneliners but about conditional assignments using the ternary operator. if/else should be used when different branches in execution are taken but here it's not really about a procedure but in the end you want to have a value assigned to `$job->{TEST}`. If you think this is not clear enough then maybe we should find a way where we don't need to replace the detected number but instead make sure that we always apply a sensible suffix

Reply:
oh, I just realized that the suffix is not selected in a good way because "-2" would be sorted after "-19". So instead you should look into a way how the numbers of necessary digits can be derived to be equivalent to the bash for loop:

```
for i in {001..100} ; …
```

which yields numbers

```
001
002
…
010
…
099
100
```

with the zeroes prefixed.

Reply:
The way it is right now looks clear enough to me, and I see no benefit over using a ternary operator instead, also maybe I'm not following the code correctly but, shouldn't `$job->{TEST} = $test_name."-$n"` be enough here? or am I missing something?

Reply:
@foursixnine If we simply append the number each iteration, we will end up with jobs named foobar-1-2-3-4-5-6 and so on :) 

My first proposal to have a fixed amount of digits was intended to obtain clean formatting with zeroes prefixed; so let's think about a way to format it properly with a variable number of digits.

Reply:
for the padding, `sprintf("%04d", $n)` should solve it, but from what you are saying, each job is a clone of the previous one and not the original one?

Reply:
I wonder if getting the test name or passing it to the `append_idx_to_test_name` is something that can be done, after all.. by the time we enter in the for loop, `post_params` is populated already by clone_job, so extracting the original name of the test is/should be possible..

Reply:
job is cloned once, and we only change the TEST name in the parameters hash by appending a number. Each iteration keeps the hash, so if we just append, it will be appended each time. As was requested to have no limit in `repeat` , maybe 04 is not enough, so we should do something like

```Perl
sub clone_jobs ($jobid, $options) {
    my $url_handler = create_url_handler($options);
    my $repeat = delete $options->{repeat} || 1;
    my $digits = length $repeat;
    clone_job($jobid, $url_handler, $options, my $post_params = {}, my $jobs = {});
    for my $counter (1 .. $repeat) {
        if ($repeat > 1) {
            my $number = sprintf("-%0${digits}d", $counter);
            append_idx_to_test_name($number, $post_params);
        }
        my $tx = post_jobs($post_params, $url_handler, $options);
        handle_tx($tx, $url_handler, $options, $jobs) if $tx;
    }
}
```

Reply:
And BTW while I understand the "no limit" argument, I still wonder how many use cases we have to clone a job more than thousands times :D

Reply:
> I wonder if getting the test name or passing it to the `append_idx_to_test_name` is something that can be done, after all.. by the time we enter in the for loop, `post_params` is populated already by clone_job, so extracting the original name of the test is/should be possible..

we should really talk about test `names`, because a clone operation can lead to parents/children/parallel jobs also getting cloned, each with its own name to modify. So we need an iteration over the nested hash 

```Perl
# append a formatted "-NN" to the TEST parameter
sub append_idx_to_test_name ($n, $post_params) {
    foreach my $job_key (keys %$post_params) {
        my $job = $post_params->{$job_key};
        my $test_name = $job->{TEST};
        ...
    }
}
```

Reply:
> And BTW while I understand the "no limit" argument, I still wonder how many use cases we have to clone a job more than 9 thousands times :D

The limit here, is me trying to workaround openQA not having api-rate limiting, nothing limits the user to circumvent this, but at least lets make them write a `for loop` or hack their way across 🙃

Reply:
> about test names for this use case a good way to start is
> 
> ```
> openqa-clone-job --skip-chained-deps --parental-inheritance --within-instance …
> ```
> 
> with the "--skip-chained-deps" no parents in upper layers, e.g. "image creation jobs", are cloned and with "--parental-inheritance" test parameters affect cloned parents+children.

The dependencies will end up or should end up with their dependencies where they should be, if the interest is 10 tries of X job, the parents only become relevant when they fail, and if, the children will have a reference. 

If somebody wishes also extra changes to the whole chain, they should use isos post, in any case, I'd leave handling of dependencies (parent jobs et al) to a separate PR.

Reply:
With the latest change, I addressed the padding issue

Reply:
> about test names for this use case a good way to start is
> 
> ```
> openqa-clone-job --skip-chained-deps --parental-inheritance --within-instance …
> ```
> 
> with the "--skip-chained-deps" no parents in upper layers, e.g. "image creation jobs", are cloned and with "--parental-inheritance" test parameters affect cloned parents+children.

Sorry but I don't understand what this has to do with the PR itself ? With the actual implementation, all cloned tests (parents, siblings and so on) gets the name numbered. If you skip cloning parents, only childrens will have the name modified.

Reply:
Yes, that's fine. I wanted to provide that users can still specify `--parental-inheritance` if they want custom test settings overridden in all created jobs.

Reply:
I still https://github.com/os-autoinst/openQA/pull/5331#discussion_r1362558824 as open. I suggest to use the ternary operator for the alternatives in assignment. As an alternative maybe we can find an approach so that we don't need to replace the value in all. How about reading "TEST" from the first job and just overwriting the test variable in the for-loop with an iterating value. WDYT?

Reply:
> I still [#5331 (comment)](https://github.com/os-autoinst/openQA/pull/5331#discussion_r1362558824) as open. I suggest to use the ternary operator for the alternatives in assignment. 

This sounds like a strawman argument over something that has zero value.

Suggesting usage of a ternary operator, over readability doesn't make much sense... otherwise I'd suggest to come up with a tidy/perlcritic rule that enforces ternary operators

Reply:
> This sounds like a strawman argument over something that has zero value.

Well, I stated it because I do see the value. Sorry to see that you think I would provide something with no value.

I wouldn't know how to put that into a perlcritic rule. And for everything that we can't or do not know how to put into a static check we have the human review process, right? So this issue here right now is the only open point that I know of. I see the following possibilities:
1. Use a ternary operator for the assignment
2. Follow my idea about reading TEST on the outer scope and iterate over that
3. Ask me to do provide a working implementation – I would be happy to try – and add to your branch
4. State explicitly that you will not change it and you definitely think this is better. Then I would need to think about.

Reply:
Can I ask a very naive question? Why can't we simply print the current iterator always without any conditionals? The bash equivalent does that.

Reply:
Thanks @okurz for the review. I already expressed my opinion on ternary operators and I would like to avoid the usage when possible, like in this case.
About the second option, it was my first implementation but it requirese to keep the "original" set of test names into an hash `{ job_key => name }` and assign them to the proper numbered value just before posting, or change it on the fly. I assumed this would complicate the code too much, because you also need to have the "original" hash in the `append_idx` function; so I came up with another proposal, which I just submitted including the fix for unit test output.  

```Perl
# append a formatted "-NN" to the TEST parameter for each job being posted
sub append_idx_to_test_name ($n, $digits, $post_params) {
    my $suffix = sprintf("-%0${digits}d", $n);
    foreach my $job_key (keys %$post_params) {
        my $job = $post_params->{$job_key};
        if ($n == 1) {
            $job->{TEST} .= $suffix;    # just append at the first time
        }
        else {
            # from the second onwards, replace old with the new number
            $job->{TEST} =~ s/-\d+$/$suffix/;
        }
    }
}
```

of course the owner will have the last word :)

Reply:
> Can I ask a very naive question? Why can't we simply print the current iterator always without any conditionals? The bash equivalent does that.

If we want to do exactly the same thing the bash for..loop does, we can just move the `clone_job` inside the `for` loop and you will get each iteration a "clean" TEST name. 
At this point, it's just a matter of appending the -00NN number, and job done. 

My proposal sounds to me like an occasion to use fewer resources, avoiding repeating N times the same `clone` operation and doing it only once, then `POST` N times the cloned job(s) data structure.


```Perl
sub clone_jobs ($jobid, $options) {
    my $url_handler = create_url_handler($options);
    my $repeat = delete $options->{repeat} || 1;
    my $digits = length $repeat;
    clone_job($jobid, $url_handler, $options, my $post_params = {}, my $jobs = {});
    for my $counter (1 .. $repeat) {
        append_idx_to_test_name($counter, $digits, $post_params) if $repeat > 1;
        my $tx = post_jobs($post_params, $url_handler, $options);
        handle_tx($tx, $url_handler, $options, $jobs) if $tx;
    }
}
```

Reply:
Ok. That sounds similar to something that @perlpunk and me have just come up with. Please take a look into https://github.com/os-autoinst/openQA/pull/5336 where I added a second commit with the refactoring on top of your branch. If you like the approach you can cherry-pick my commit https://github.com/os-autoinst/openQA/pull/5336/commits/6341433ee715ccdde940baf97adc02c4c87e04b8 into your branch or we merge mine.

Reply:
But `clone_job` is actually doing some requests to get the dependencies, so I think @ilmanzo is right that this is maybe using too many resources. I wouldn't care about 100 loops of some perl code, but if it is doing 100 requests, then it's not that nice.

Reply:
NP, feel free to merge yours; thanks again because I learned a lot from this exchange of views!

Reply:
@perlpunk Yes, my proposal is to avoid what the bash loop does:

```
- clone
- post
- clone
- post
- clone
- post
- clone
- post
...
```
but do a more efficient  
```
- clone
- post
- post 
- post
- post
... 
```
even at the cost of the code being (slightly) more complex :)

Reply:
said in another way, for a `--repeat N` this implementation will spare N-1 `GET` calls to the API server, which are needed to download job dependencies on each clone operation.

Reply:
Got it. I plan to follow up with rework after that while of course keeping the approach regarding avoiding unneeded requests. At the very least we should avoid rename functions as we have identified in #5336 . That of course is something that was already in the codebase so not at all any critic about this PR here